### PR TITLE
reduce cs_main scope, guard block index 'nFile' under a local mutex

### PR DESCRIPTION
**Pull request by furszy**
By not having to lock `cs_main` every time we try to access block
position on disk, we can avoid slowing down concurrent tasks due
to threads waiting for the global lock to be released.

Also, solves a edge case race inside `ReadBlockFromDisk` where
we obtain the file position locking `cs_main` and then try to access
the file without it (the file could no longer exist if pruning occurred
just in-between these two actions).

Context from where this comes from #26966.

### CODE REVIEW

File: src/chain.h [hunk 1]
```diff
@@ -162,14 +162,16 @@ class CBlockIndex
     //! height of the entry in the chain. The genesis block has height 0
     int nHeight{0};
 
+    const std::shared_ptr<Mutex> cs_data = std::make_shared<Mutex>();
```
Comment (MarcoFalke) [inline src/chain.h:165]: Huh? Does this create a mutex for every block? What is the memory footprint?

Comment (furszy) [reply src/chain.h:165]: yeah, good topic. I was so focused on accessing disk concurrently that forgot that we have 800k block indexes loaded in memory all the time.
The mutex size is 64 bytes, so.. a mutex per block index is a no go.

Could move it to a single global mutex for block indexes file data access, which would still decouple it from `cs_main` (which is a good step forward) but.. it's still not the best solution for concurrent access.

Comment (furszy) [reply src/chain.h:165]: Working on it 👌🏼

Comment (Sjors) [reply src/chain.h]: ^ is this done, or still in progress?

Comment (furszy) [reply src/chain.h]: > ^ is this done, or still in progress?

Yeah, have pushed the readers/writer lock in the last update.

Still, have to say that code elegancy wise, I'm not totally happy with the `sync.h` shared lock wrapper. Planning to give it another look later this week.

Comment (jamesob) [reply src/chain.h]: Why not do a mutex per blockfile (rather than CBlockIndex entry, which would indeed be crazy)? There are hundreds to thousands of those, so keeping a map of blockfile number to mutex shouldn't be unreasonable.

Comment (furszy) [reply src/chain.h]: While that would be better than the initial per `CBlockIndex` entry mutex, it's not as good as a single RW lock (which is the current PR state).

Essentially, we aim to allow concurrent access for read-only operations, while write operations that are not performed regularly require exclusive access (which is the RW lock definition).

To place it more in our code. The only time where the `CBlockIndex` file position can be modified during the entire software lifecycle is pruning, and it's done only once. And because of that single specific scenario that does not happen often, we have been forced to lock `cs_main` every time that any process wants to access the block index file position.

So, guarding this single specific case with an exclusive lock, the rest of the read-only operations can be processed concurrently by sharing ownership of the readers-only lock.

Comment (furszy) [reply src/chain.h]: A little add to my last comment:

I think that if we ever add other use cases that modify the block index file position more frequently (like re-ordering/compacting block data on disk or re-downloading blocks), then a more fine-grained RW lock per block file would be a nice idea.

-- the comment came to my mind because one of my long-term projects introduces the pruned blocks re-download feature.. --

File: src/node/blockstorage.cpp [hunk 2]
```diff
@@ -127,7 +127,7 @@ void BlockManager::PruneOneBlockFile(const int fileNumber)
         if (pindex->nFile == fileNumber) {
             pindex->nStatus &= ~BLOCK_HAVE_DATA;
             pindex->nStatus &= ~BLOCK_HAVE_UNDO;
-            pindex->nFile = 0;
+            pindex->nFile = -1;
```
Comment (Sjors) [inline src/node/blockstorage.cpp]: 47bc99cddc518dc41741c5dcc17f9e629e162449 This is probably fine, but there's a bunch a places that we have to make sure are never reached. For example in `FindBlockPos(` we do `_blockfile_info[nFile]` where `nFile = ... pos.nFile`. I'll have to review that a bit more.

File: src/kernel/chain.cpp [hunk 3]
```diff
@@ -16,9 +16,8 @@ interfaces::BlockInfo MakeBlockInfo(const CBlockIndex* index, const CBlock* data
     if (index) {
         info.prev_hash = index->pprev ? index->pprev->phashBlock : nullptr;
         info.height = index->nHeight;
-        LOCK(::cs_main);
-        info.file_number = index->nFile;
-        info.data_pos = index->nDataPos;
+        info.file_number = index->GetFileNum();
+        info.data_pos = index->GetDataPos();
```
Comment (LarryRuane) [inline src/kernel/chain.cpp:20]: Shouldn't these be fetched atomically? Maybe there should be a getter that returns all three values, similar to how there's a setter (`CBlockIndex::SetFileData`) than sets all three atomically.

**Review by stickies-v — COMMENTED** [review]
Concept ACK

Is my understanding correct that 1) replacing `cs_main` with `g_cs_blockindex_data` and 2) having `g_cs_blockindex_data` be a shared instead of recursive/exclusive mutex are orthogonal improvements? I think it makes sense for both of them to happen in this PR, just wondering.

File: src/kernel/chain.cpp (hunk 3 above)
Comment (stickies-v) [reply src/kernel/chain.cpp:20]: Starting to feel like a struct for those 3 members would be appropriate

Comment (furszy) [reply src/kernel/chain.cpp:20]: > Shouldn't these be fetched atomically? Maybe there should be a getter that returns all three values, similar to how there's a setter (`CBlockIndex::SetFileData`) than sets all three atomically.

@LarryRuane I didn't add it merely to not introduce that many changes but could be done too. The members set is guarded by `cs_main` (for now) so there is no possible race.

> Starting to feel like a struct for those 3 members would be appropriate

Why exactly?. So far in the code, we either want the block data position or the block undo data position, we don't get both at the same time because they represent different files on disk (thus why they are returned inside `FlatFilePos`, the struct groups only one of them with the file number).

In other words, nData + nFile maps to a "blk<nFile>.dat", while nUndoData + nFile maps to an specific "rev<nFile>.dat".

**Comment by furszy** [comment]
> Concept ACK
> 
> Is my understanding correct that 1) replacing `cs_main` with `g_cs_blockindex_data` and 2) having `g_cs_blockindex_data` be a shared instead of recursive/exclusive mutex are orthogonal improvements? I think it makes sense for both of them to happen in this PR, just wondering.

Moving `cs_main` to an blockstorage local exclusive mutex would still be an improvement from the current state, yeah. The networking threads (and pretty much the entire node) shouldn't get blocked because the user called the `getblock` RPC command.

But, the usage of an exclusive mutex here still means no concurrent block data access. Which is a shame as accessing block 124 on disk shouldn't block the entire node from accessing block 750,000 on disk (nor any other block). They are on different files.

File: src/kernel/chain.cpp (hunk 3 above)
Comment (LarryRuane) [reply src/kernel/chain.cpp:20]: Are you sure these are protected by `cs_main`, is it? Running `gdb test_bitcoin` on the PR branch (leaving some irrelevant stuff out):
```
(gdb) b MakeBlockInfo
(gdb) run
Thread 29 "b-basic block f" hit Breakpoint 1, kernel::MakeBlockInfo (index=0x55555749b018, data=0x0) at kernel/chain.cpp:14
#0  kernel::MakeBlockInfo (index=0x55555749b018, data=0x0) at kernel/chain.cpp:14
#1  0x000055555623fa81 in BaseIndex::ThreadSync (this=0x7fffffffad00) at index/base.cpp:201
(gdb) p cs_main
$1 = {
  <std::recursive_mutex> = {
    <std::__recursive_mutex_base> = {
      _M_mutex = {
        __data = {
          __lock = 0,
          __count = 0,
          __owner = 0,
          }
        }
      }
    }
  }
```
I agree with you that a new struct should just be `nFile` and ` nDataPos` (or `nPos`). This struct can be used for either block data or rev data.

**Review by dergoegge — CHANGES_REQUESTED** [review]
Concept ACK on reducing `cs_main` scope.

Approach NACK for adding new globals. It is not worth going with a worse design to "rush in" an improvement for a feature that may or may not happen (#26966). I would prefer to go with a more sustainable approach from the get go, otherwise this will need to be refactored in the long run.

I don't see why the mutex could not be a member of `BlockManager`. Like you say in https://github.com/bitcoin/bitcoin/pull/27006#issuecomment-1481222148, giving `BlockManager` its own internal lock(s) instead of `cs_main` is the preferable long term design.

File: src/kernel/chain.cpp (hunk 3 above)
Comment (furszy) [reply src/kernel/chain.cpp:20]: > Are you sure these are protected by cs_main, is it?

The members' write process, not the read. If the read would be protected by `cs_main`, this PR wouldn't make any sense :p.
The two places are: the `CBlockIndex` creation time and `CBlockIndex` pruning time.

Still, those `cs_main` locks can also be removed in a follow-up PR. I haven't removed them here because they require further changes.

Not sure if I'm understanding your concern here. You are thinking on a race in-between `GetFileNum` and `GetDataPos`?
Because, if that is the concern, the error that could arise from it is harmless. The read block call would just return false, and callers would continue. But, to not even have to think about such error, we could just call `GetFilePos()` instead of the two getters.

> I agree with you that a new struct should just be nFile and  nDataPos (or nPos). This struct can be used for either block data or rev data.

There is already a method for that: `CBlockIndex::GetFilePos`.

**Comment by dergoegge** [comment]
> I never said the opposite. Not sure why the "rush" term is being used here really. I haven't said anything like "this is great, we must have it now".. (nor here nor anywhere in the repository).

I'm sorry didn't mean to put words in your mouth. It seemed rushed to me because you are going with an "easier" solution here rather than the more complex but (imo) worth while long term design.

> it will need blockstorage dependency to use the thread analysis annotation

I was thinking of not adding the annotations in this case. Which is not great but seems better to me than adding more globals.

> Or.. thinking out loud: something somewhat "radical" could be move nFile, nData and nUndoData members out of CBlockIndex, into the blockstorage class. We could place them into an unordered map <block_id, FileData>.

This separates the concerns nicely and lets us add annotations, I would be happy with this :) The only thing that's a bit weird then is maintaining the on-disk format for the block index entries but that seems workable.

**Comment by furszy** [comment]
> > I never said the opposite. Not sure why the "rush" term is being used here really. I haven't said anything like "this is great, we must have it now".. (nor here nor anywhere in the repository).
> 
> I'm sorry didn't mean to put words in your mouth. It seemed rushed to me because you are going with an "easier" solution here rather than the more complex but (imo) worth while long term design.

Well, this PR didn't have any long term design goal discussion prior to your arrival.
I'm happy that it happened as it pushed us to think about the "radical" change mentioned above.

> > it will need blockstorage dependency to use the thread analysis annotation
> 
> I was thinking of not adding the annotations in this case. Which is not great but seems better to me than adding more globals.

I'm on the other side of the fence actually. While globals aren't ideal, particularly for testing, I prefer them over loosing thread safety analysis as it is essential to prevent deadlocks.

> > Or.. thinking out loud: something somewhat "radical" could be move nFile, nData and nUndoData members out of CBlockIndex, into the blockstorage class. We could place them into an unordered map <block_id, FileData>.
> 
> This separates the concerns nicely and lets us add annotations, I would be happy with this :) The only thing that's a bit weird then is maintaining the on-disk format for the block index entries but that seems workable.

Yeah, the read shouldn’t be that hard. And a db upgrade mechanism is doable too.

**Comment by furszy** [comment]
Closing for now. Might revive it in the future, post #26966.
